### ISSUE TITLE
How do I change the storage directory?

### ISSUE THREAD

**Issue by Prillan** [input-output-hk/daedalus#605]
I don't want it to consume my remaining disk space on my SSD, how can I change the storage directory to something other than `%APPDATA%`?

This is actually preventing me from trying Cardano.

**Comment by smukherj1** [comment]
I found a workaround if you are using the Windows 64-bit client. Note: This is a workaround and it involves tinkering with the launcher command. Please don't hold me accountable if you lose your tokens or any other data.

First turn off Daedalus if it's running.
Go to your Daedalus installation directory and look for the file called "daedalus.bat". In the line after
`-n --db-path ^`
change 
`-n "%APPDATA%\Daedalus\DB Wallet 1.0" ^`
to wherever you want to store your blocks. 

Start Daedalus. It should start downloading blocks from 0% to the new directory.

Obviously this is an ugly hack and can break with a new Daedalus version but it works for 1.0.3769 Windows 64-bit.

**Comment by cleverca22** [comment]
the automatic updates will also revert such a change when they happen

**Comment by CryptonZylog** [comment]
A workaround I found is to create a NTFS directory link to another disk. Backup your appdata directory of course, then follow [instructions](https://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/) to create your own.

**Comment by smukherj1** [comment]
Yeah I like the symlink idea better than mine :) Would really prefer an official solution though in case updates also change the names of data directories

**Comment by CryptonZylog** [comment]
#595 seems to be a similar problem.

**Comment by realuser** [comment]
+1:"This is actually preventing me from trying Cardano."

**Comment by snowkidind** [comment]
+10 

i have tried to get daedalus installed on my mac pro with a 256 gig hard drive that costs 1500 dollars to replace with a larger one. In the ten free gigs i have on that machine, daedalus didnt even get the entire blockchain loaded before it filled it up.

I have 40 hard drives but i cannot point daedalus to any of them because it links to stupid system drive

this is a real issue, please fix it.

**Comment by xjmzx** [comment]
Yes. Managing data on one's own machine is both expected and desirable. Any thoughts from deveopers would be appreciated

**Comment by priyank8** [comment]
While I understand the need of the location for blockcain data to be user selectable, the private wallet needs to be in user profile directory to keep it secure with multiple users that can log on to a system. For short term, you can create a simple batch/shell script (windows example below):
1. just create a new file and call it 'daedalus.bat'
2. open it in notepad and paste below (replace <target-dir> with desired path):
set LocalAppData=D:\<target-dir>
cd /D "C:\Program Files\Daedalus"
cardano-launcher.exe
3. save the file
4. double click on it

Note that this would mean your profile folder for Daedalus is created in D: , you might want to restrict permissions to the folder

**Comment by priyank8** [comment]
While I understand the need of the location for blockcain data to be user selectable, the private wallet needs to be in user profile directory to keep it secure with multiple users that can log on to a system. For short term, you can create a simple batch/shell script (windows example below):

just create a new file and call it 'daedalus.bat'
open it in notepad and paste below (replace with desired path):
set AppData=D:<target-dir>
cd /D "C:\Program Files\Daedalus"
cardano-launcher.exe
save the file
double click on it
Note that this would mean your profile folder for Daedalus is created in D: , you might want to restrict permissions to the folder

**Comment by weaver299** [comment]
I've been using the Daedalus wallet on a Macbook, but the blockchain is taking too much storage on the SSD.  I'd like to move it to an external drive.

I've just been trying Cardano out, and am not significantly personally invested in it.  I might just transfer my ADA back to Binance and sell it if I have to commit all this SSD space to the blockchain...

thanks for listening...

**Comment by priyank8** [comment]
>if I have to commit all this SSD space to the blockchain...
You have multiple options now.
If you dont need to run a full node, you can simply switch to one of the light wallets:
https://cardanowiki.info/wiki/Comparison_of_Ada_wallets_and_exchanges

**Comment by nikolaglumac** [comment]
This feature is not yet supported but there are plans to add support in the future.

**Comment by weaver299** [comment]
Why are the issues being closed if this feature is not implemented, but is planned for the future?
Is there another, canonical, issue for this?

**Comment by weaver299** [comment]
I see other issue tickets tagged "feature-request" which are not closed... should this not be treated the same as those?

**Comment by nikolaglumac** [comment]
Yes, you are right - sorry about that - there were many duplicates and I forgot to leave one open.
Will leave your's open 👍

**Comment by weaver299** [comment]
Okay, thanks.. I feel better now. :)

**Comment by nikolaglumac** [comment]
Fear not - even if the issue has remained closed we would not forget about this feature request :)
It is on our todo list!

**Comment by BrendanJ** [comment]
Any ETA?

**Comment by nikolaglumac** [comment]
Hi @BrendanJ,

Unfortunately, there is no clear date at the moment.
The team is fully focused on the upcoming Byron Reboot and Shelley releases.

Thank you for your patience and understanding!

**Comment by garyhussey** [comment]
Is this on the roadmap yet?  Cardano is about to go Mainnet and this is still a problem.  I mean... most people can manage a config file... just move the damn variable somewhere we can change it - config.json perhaps?  We're not asking for an interface... just the option to put our database somewhere it's not going to eat primary resources.

**Comment by rdlrt** [comment]
@nikolaglumac - Unfortunately, I have to agree, not sure why a basic option as this isnt part of installer prompt to select (users are only talking about blockchain directory - previously DB-1.0, now chain, not wallet folders) is not doable in more than 25 months. Suggesting workarounds feels suboptimal.

@garyhussey - the workaround users have been using is on forum [here](https://forum.cardano.org/t/storing-the-cardano-blockchain-on-another-hard-drive-flash-drive/3699) (via symlinks), tho note that the profile folder name is now `Daedalus Mainnet` instead of `Daedalus`.

**Comment by nikolaglumac** [comment]
@disassembler can you help us out here please?

**Comment by garyhussey** [comment]
@rdlrt, thank you for the link. I've searched all over the place for something other than the symlink + re-installation method.

**Comment by rdlrt** [comment]
Hi @garyhussey ,
Only reason I dont suggest updating config files is an update could overwrite them and add duplication of directories. Until its an option available natively, using a symlink or custom profile keeps it available across updates

**Comment by garyhussey** [comment]
@rdlrt, I agree for any configuration files in the app root folder, but the config.json file is in the APPDATA directory.  I'd assumed that it was user specific settings that were not over-written by updates.  I could be wrong, I've not had an opportunity to make myself familiar with the build process yet.

**Comment by wrapperband** [comment]
Please - can you fix this very serious issue, it's being going on a long time and causes users extreme time effort. It surely is trivial to add changing the database directory, or should be....... It is not right to use excuses like use a light weight wallet. If you want us to use symlink - make the help say how to do it (in linux).
I am now stuck with no space on my main disk, because Daedelus only installs on home disk. This may have already caused my data loss and has already caused me a number of hours work trying to change the directory like most other main crypto wallets.

**Comment by wrapperband** [comment]
Just as an Update - I've tested the symlink fix and was finally able to create a link to the copied mainnet directory- 

 - mainnet -> /media/HomeData/.local/share/Daedalus/mainnet

Not an expert on symlinks but the Daedalus program does not now start with a symlink in Ubuntu/Linux.

**Comment by KudretBayram** [comment]
Just writing to add a little more weight to this. It's a real issue, let's get it done please.

**Comment by nikolaglumac** [comment]
@KudretBayram we will do our best to get this implemented as soon as possible. Thank you for your patience! 🙏

**Comment by nikolaglumac** [comment]
This feature is not yet implemented but this is a workaround that might help you:

1. Close Daedalus
2. Open the state directory: Instructions
3. Copy the "chain" directory to a different drive
4. Create a shortcut with the same name "chain" to that new location and place it in the state directory (where "chain" directory was previously)
5. Run Daedalus

This will free your C drive significantly since the "chain" directory makes 90% of state directory size.

**Comment by KudretBayram** [comment]
Thanks Nicola, I appreciate your time in responding to this thread.
Good idea. I tried it (on a MAC) and received the following error:

"Cardano Node Crashed"

**Comment by nikolaglumac** [comment]
@gnpf @ManusMcCole @mioriohk can you please try this out and post results for all 3 platforms please 🙏

**Comment by mioriohk** [comment]
Hello everyone. Hope everybody is well 👍 

In order to create symlink on Daedalus chain folder, for all three platforms (`Windows, Mac and Linux`), please follow these steps:
`Note: Replace <username> with your own username`


1. Stop Daedalus

2. Find the chain folder according to your chosen platform. Normally located at:
    For Windows: 
   ` "C:\Users\<username>\AppData\Roaming\Daedalus Mainnet\chain" `

    For Mac:
    `“/Users/<username>/Library/Application\ Support/Daedalus\ Mainnet/“`

    For Linux:
    `“/home/<username>/.local/share/Daedalus/mainnet”`

 3. Then move the “`chain`” folder and its entire content to your desired location. I moved mine into “`Downloads`” folder.

 4. After that do the following:
     For Windows:
         a) Run the command prompt as an administrator
         b) At the prompt, type -> `mklink /d "C:\Users\<username>\AppData\Roaming\Daedalus Mainnet\chain" "C:\Users\<username>\Downloads\chain"`

     For Mac:
         a) Open a terminal window
         b) At the prompt, type -> `ln -s /Users/<username>/Downloads/chain /Users/<username>/Library/Application\ Support/Daedalus\ Mainnet/`

     For Linux:
         a) Open a terminal window
         b) At the prompt, type -> `ln -s /home/<username>/Downloads/chain /home/<username>/.local/share/Daedalus/mainnet`

   5. Restart Daedalus

If you have any question, please let me know. Thank you. 🙏

**Comment by KudretBayram** [comment]
This appears to have worked!
Thank you.

**Comment by Morgorth** [comment]
there is now an option to change it. 
Daedalus Diagnostic > Daedalus State Directory

**Comment by nikolaglumac** [comment]
@Morgorth that is a view only option. You can **not** change that value.

**Comment by CantTouchDis** [comment]
I tried to change the chain directory in the same way miorsufianiohk explained in: [his post](https://github.com/input-output-hk/daedalus/issues/605#issuecomment-771782614) but it did not work as expected. What happened instead is that a new directory with the path of the link was created in my `/tmp` folder. (I assume this is caused by the path I linked to being on a different SSD)

After that I just changed the home volume in the exec command in `.local/bin/daedalus-mainnet`  from 
```
... -m /home:/home ...
```
to
```
... -m [MyDaedalusDirectory]:/home ...
```
This worked for me

**Comment by Sebastix** [comment]
> [MyDaedalusDirectory]

Same happens here (Linux - Lubuntu OS).
@CantTouchDis Can you give an example of a path you used for `[MyDaedalusDirectory]` ?

**Comment by CantTouchDis** [comment]
In my case it was something like `/media/Samsung1TB/daedalus` but it could be any path

**Comment by greg-fischer** [comment]
Ok so I first thought this was working, but it was causing a resync of the chain after moving and symlinking, which is annoying on a slow connection.  But after checking files in use with "lsof", its not even touching the external filesystem I link to.  It's creating it in /tmp!  The symlink is pointing to the correct destination.   I'm on Ubuntu, btw.  

This should be configurable to the user, even if in a config file.
It should also not force a resync from scratch when moved.
And... if the software is ignoring links in anyway, stop it!  Don't translate that path!  If I want to symlink, honor it!  

Sorry if I'm a little grumpy. I keep running into software lately that ignore symlinks and get the canonical path to use instead, which is extremely frustrating for linux users who have been managing things like this for years.

I don't have room on /root or /home for multiple blockchains of 10's for GB's in size.  

Does Daedalus ask for install destination on Linux?  I can't remember if it did, but I could just reinstall if that would work.

**Comment by weaver299** [comment]
Agree with @greg-fischer ... this is a deeply frustrating situation and is honestly really undermining my confidence in the project, if something like _where the blockchain data is stored_ can't be configured.  This seems like a minimum-viable-product feature... if this hasn't been implemented for _years_, what does that say about the software quality of this project.

**Comment by CantTouchDis** [comment]
@greg-fischer have you tried my solution from https://github.com/input-output-hk/daedalus/issues/605#issuecomment-794107731 ? this worked for me and I put it on a different ssd

**Comment by greg-fischer** [comment]
> @greg-fischer have you tried my solution from [#605 (comment)](https://github.com/input-output-hk/daedalus/issues/605#issuecomment-794107731) ? this worked for me and I put it on a different ssd

OH I see!  That's what I was looking for.   I'll test that later.  I read your post before but probably skimmed too quickly because I just wasn't following it. (until now :) 
thanks!  I'll report back!

**Comment by greg-fischer** [comment]
Sorry @CantTouchDis but this is not working for me.  It always runs from /tmp.  I tried doing the symlink for just the chain folder.  Then tried several different options for the -m /home:/home option, it always runs in /tmp.

I have an external home folder: /home-ext/greg 
(my $HOME is /home/greg - the above is just for things like a blockchain that's huge and don't need backed up!)
In here I created/rsynced the same .local/share/Daedalus/mainnet/* directories, then I linked just the chain folder.
ln -s /home-ext/greg/.local/share/Daedalus/mainnet/chain /home/greg/.local/share/Daedalus/mainnet/chain

The symlink is there and works correctly.  I then changed the file: /home/greg/.local/bin/daedalus-mainnet
Changed -m /home:/home to -m /home-ext:/home

Still goes to /tmp.  This is all after I removed Daedalus and installed fresh again.

I tried several options for /home:
-m /home-ext/greg:/home
-m /home-ext/greg/daedalus:/home
-m /home-ext:/home
-m /home-ext/greg/.local/share/Daedalus/mainnet:/home

I know you posted that nix looks for that as your home folder, but I assume it appends your username, so I think this would be the right option:
-m /home-ext:/home

I have full permission in /home-ext/greg.

One thing I am just now wondering is if Daedalus needs to be installed with sudo or not. I installed as myself, no sudo.  Didn't get any errors and it launches.

**Comment by nikolaglumac** [comment]
@dmitrii-gaico @miorsufianiohk can you please assist us here?

**Comment by miorsufianiohk** [comment]
Hi @greg-fischer ,

I hope that you are well. Could I check that you have followed the [instruction written here](https://github.com/input-output-hk/daedalus/issues/605#issuecomment-771782614)?

Kind regards,
Mior

**Comment by greg-fischer** [comment]
Thanks for the help @miorsufianiohk 

I just read through that.  

Is this the "chain" folder to symlink?
“/home/<username>/.local/share/Daedalus/mainnet”
Or is this it?
“/home/<username>/.local/share/Daedalus/mainnet/chain”


So I tried this first:
“/home/greg/.local/share/Daedalus/mainnet” 
is a symlink to > “/home-ext/greg/daedalus/mainnet”

```
greg@thardus:~$ cd .local/share/Daedalus
greg@thardus:~/.local/share/Daedalus$ ls -la
total 4
drwxrwxr-x  2 greg greg   21 Jun 29 03:04 .
drwx------ 21 greg greg 4096 Jun 29 03:01 ..
lrwxrwxrwx  1 greg greg   31 Jun 29 03:04 mainnet -> /home-ext/greg/daedalus/mainnet
greg@thardus:~/.local/share/Daedalus$ ls mainnet/
cardano-node.socket  chain  daedalus_lockfile  icon_large.png  icon.png  Logs  namespaceHelper  Secrets  tls  wallets
greg@thardus:~/.local/share/Daedalus$ ls /home-ext/greg/daedalus/mainnet
cardano-node.socket  chain  daedalus_lockfile  icon_large.png  icon.png  Logs  namespaceHelper  Secrets  tls  wallets
greg@thardus:~/.local/share/Daedalus$ 
```
Now Daedalus will not open. It starts to launch and opens an icon on the dock, then it's gone.


Then I tried linking the "chain" folder inside mainnet, which is what I've been doing all along. 

```
greg@thardus:~/.local/share/Daedalus/mainnet$ ls -la
total 596
drwxrwxr-x 6 greg greg    185 Jun 29 03:29 .
drwxrwxr-x 3 greg greg     21 Jun 29 03:27 ..
srwxrwxr-x 1 greg greg      0 Jun 29 03:29 cardano-node.socket
lrwxrwxrwx 1 greg greg     29 Jun 29 03:28 chain -> /home-ext/greg/daedalus/chain
-rw-rw-r-- 1 greg greg      0 Jun 29 03:01 daedalus_lockfile
-rw-r--r-- 1 greg greg 594590 Jun 29 02:53 icon_large.png
-rw-r--r-- 1 greg greg   5641 Jun 29 02:53 icon.png
drwxrwxr-x 3 greg greg     17 Jun 29 02:53 Logs
-rwxr-xr-x 1 greg greg    880 Jun 29 02:53 namespaceHelper
drwxrwxr-x 2 greg greg      6 Jun 29 03:01 Secrets
drwxrwxr-x 4 greg greg     34 Jun 29 03:01 tls
drwxrwxr-x 2 greg greg     92 Jun 29 03:29 wallets
greg@thardus:~/.local/share/Daedalus/mainnet$ ls chain
clean  immutable  ledger  lock  protocolMagicId  volatile
greg@thardus:~/.local/share/Daedalus/mainnet$ ls /home-ext/greg/daedalus/chain
clean  immutable  ledger  lock  protocolMagicId  volatile
greg@thardus:~/.local/share/Daedalus/mainnet$ 
```

Dadalus opens fine, but its not using that linked folder. Its using /tmp.

`lsof |grep /home-ext/greg/daedalus/chain`

Outputs a ton of files used here:
`/tmp/nixlLYqq7/home-ext/greg/daedalus/chain`

```
greg@thardus:~/.local/share/Daedalus/mainnet$ ls -lasth /tmp/nixlLYqq7/home-ext/greg/daedalus/chain
total 8.0K
4.0K drwxrwxr-x 2 greg greg 4.0K Jun 29 03:35 volatile
   0 drwxrwxr-x 2 greg greg  321 Jun 29 03:34 immutable
   0 drwxrwxr-x 2 greg greg   18 Jun 29 03:29 ledger
   0 drwxrwxr-x 5 greg greg   88 Jun 29 03:29 .
   0 drwxrwxr-x 3 greg greg   19 Jun 29 03:29 ..
   0 -rw-rw-r-- 1 greg greg    0 Jun 29 03:29 lock
4.0K -rw-rw-r-- 1 greg greg    9 Jun 29 03:29 protocolMagicId
greg@thardus:~/.local/share/Daedalus/mainnet$ ls -lasth /home-ext/greg/daedalus/chain
total 62K
 11K drwxrwxr-x 3 greg greg  3 Jun 29 03:28 ..
 11K drwxrwxr-x 5 greg greg  8 Jun 29 03:01 .
 512 -rw-rw-r-- 1 greg greg  0 Jun 29 03:01 clean
 11K drwxrwxr-x 2 greg greg 13 Jun 29 03:01 volatile
 11K drwxrwxr-x 2 greg greg  3 Jun 29 03:01 ledger
 11K drwxrwxr-x 2 greg greg  5 Jun 29 03:01 immutable
 512 -rw-rw-r-- 1 greg greg  0 Jun 29 03:01 lock
6.0K -rw-rw-r-- 1 greg greg  9 Jun 29 03:01 protocolMagicId

```



Your help is greatly appreciated.   If you have any other suggestions, I'll have to look again tomorrow night. :) Thanks again!

**Comment by miorsufianiohk** [comment]
Hi @greg-fischer ,

The folder that you need to symlink is this one: /home/<username>/.local/share/Daedalus/mainnet/chain

**Comment by miorsufianiohk** [comment]
Hi @greg-fischer ,

I have been wondering if the formats of your harddrive and the external one are  the same? One thing that you could try (if you haven't yet) is [this](https://github.com/input-output-hk/daedalus/issues/605#issuecomment-353238443) .

**Comment by CantTouchDis** [comment]
@greg-fischer 
here are the contents of my `$HOME/.local/bin/daedalus-mainnet`:
```sh
...
if [ "x$DEBUG_SHELL" == x ]; then
  exec ./nix/store/apzfc51a2aki7xdyzifldp1nm4w8a1am-nix-user-chroot-2c52b5f/bin/nix-user-chroot -n ./nix -c -e -m /media/Samsung1TB/daedalus:/home -m /etc:/host-etc -m etc:/etc -p DISPLAY -p HOME -p XAUTHORITY -p LANG -p LANGUAGE -p LC_ALL -p LC_MESSAGES -- /nix/var/nix/profiles/profile-mainnet/bin/enter-phase2 daedalus
else
...
```
not sure why this branch is used, but maybe you need to change both the if and else branches occurrences

`/media/Samsung1TB/` is the mount point of my second SSD both are `ext4`.

I did not create any links in `.local/share/daedalus`, because this folder does not exist for me.

Instead of this folder Deadalus uses `/media/Samsung1TB/daedalus/`.

In this folder daedalus created a folder for my user:
```
bauschp@mycomputer:~$ ls -la /media/Samsung1TB/daedalus/
total 489412
drwxr-xr-x  3 bauschp users        4096 Mar  9 17:05 .
drwxr-xr-x 12 bauschp bauschp      4096 Jun  5 20:00 ..
drwxr-xr-x  7 bauschp users        4096 Mar 31 13:30 bauschp
-rwxr-xr-x  1 bauschp users   394194987 Mar  6 12:26 daedalus-3.3.2-mainnet-16919.bin
```

with all the contents in it:

```
bauschp@mycomputer:~$ ls /media/Samsung1TB/daedalus/bauschp/.local/share/Daedalus/mainnet/
chain              Downloads       icon.png  namespaceHelper  Secrets  wallets
daedalus_lockfile  icon_large.png  Logs      safemode.yaml    tls
```

Hope this helps, maybe the different filesystems cause problems for you, you could also check if it makes a difference if you have a separate folder for your "fake" nix home path. (`/home-ext/myFolder` instead of `/home-ext/`)

Good luck, keep us informed if anything comes up

**Comment by greg-fischer** [comment]
OK! Solved!

First note though, ZFS isn't the issue. It was working on ZFS on another system, I realized later.

The problem is that there's two different scripts that start Daedalus! One CLI, another for applications menu.  When I run the application from the Ubuntu menu it was ignoring what I put in ~/.local/bin/daedalus-mainnet.  Now that I've slept a bit, it occurred to me to run that script from the CLI directly, which worked and used my /home-ext/greg/daedalus folder.   So then I realized, "OK, what's the menu launcher running then?"   A totally different launcher script of course.  I should have questioned this before.

My desktop menu launcher is in this file:
`~/.local/share/applications/Daedalus-mainnet.desktop`

In there it shows the Exec line to the script to launch Daedalus:
`/home/greg/.local/share/Daedalus/mainnet/namespaceHelper`

That script appears to be the same as ~/.local/bin/daedalus-mainnet script.  I changed the -m /home:/home to match my folder and now it works.  I don't think there's a need for symlinking with this method.  

This brings up another issue though. It looks like changing this -m option runs the entire wallet from the new location.  This is not what I'd prefer.  It's ok for now, solved my immediate problem.  However, I want my wallet in my $HOME and the massive public blockchain somewhere else.  IMO, it would be best to symlink the "chain" folder elsewhere, but something in Nix or maybe the Daedalus app itself thinks /tmp is a better idea.  Actually, this really should be in the software as a configurable option, or config file at the very least.

I also want to point out all these blockchain wallets storing a copy of the chain in users $HOME's is horrible.  I'll have over 30G of public blockchain data that's pointless to backup between a few wallets.  User's typically want to backup their home folder, and if most people run backups with this configuration, that's a huge waste of resources.  Especially if one backs up to cloud... what a waste of time, bandwidth, and storage.

**Comment by CantTouchDis** [comment]
ok good to know.

I agree with you as I mentioned above

> I would just use that as a workaround for now until there is an option to store the chain in a separate directory

An option to change the chain directory would be a nice addition

**Comment by ghosty111** [comment]
Almost 4 years and no change how about that ....

**Comment by InderKumarRathore** [comment]
@ghosty111 if you read the comments you will find the solution to do it, it's not convenient but it works.

**Comment by nikolaglumac** [comment]
> Almost 4 years and no change how about that ....

Sorry to disappoint you... As @InderKumarRathore mentioned - we hope to get to this once Alonzo work is done. Thanks for the patience! 🙏

**Comment by danger89** [comment]
Thanks for rolling-out smart contracts rather soon than later on the mainnet, so you can update Daedalus with those must needed features indeed!

**Comment by ancoron** [comment]
Here is my take on it (a bit more generic). I just modified the `~/.local/bin/daedalus-mainnet` script into this:
```bash
#!/usr/bin/env bash

set -e

cd ~/.daedalus/
mkdir -p etc
cat /etc/hosts > etc/hosts
cat /etc/nsswitch.conf > etc/nsswitch.conf
cat /etc/localtime > etc/localtime
cat /etc/machine-id > etc/machine-id
cat /etc/resolv.conf > etc/resolv.conf

nix_opts=()

if [ -d "${DAEDALUS_DIR}" ]; then
    # mount custom data dir
    nix_opts+=(-m "${DAEDALUS_DIR}:${HOME}/.local/share/Daedalus")
fi

if [ "x$DEBUG_SHELL" == x ]; then
  exec ./nix/store/s74k20bgs8bm8267mifzxsmkgm2n94hb-nix-user-chroot-2c52b5f/bin/nix-user-chroot -n ./nix -c -e -m /home:/home -m /etc:/host-etc -m etc:/etc "${nix_opts[@]}" -p DISPLAY -p HOME -p XAUTHORITY -p LANG -p LANGUAGE -p LC_ALL -p LC_MESSAGES -- /nix/var/nix/profiles/profile-mainnet/bin/enter-phase2 daedalus
else
  exec ./nix/store/s74k20bgs8bm8267mifzxsmkgm2n94hb-nix-user-chroot-2c52b5f/bin/nix-user-chroot -n ./nix -c -e -m /home:/home -m /etc:/host-etc -m etc:/etc "${nix_opts[@]}" -p DISPLAY -p HOME -p XAUTHORITY -p LANG -p LANGUAGE -p LC_ALL -p LC_MESSAGES -- /nix/var/nix/profiles/profile-mainnet/bin/enter-phase2 bash
fi
```
Now, the only thing that needs to be set is the environment variable `DAEDALUS_DIR` (which is also used inside the nix chroot, but will not collide here). So I have this in my `~/.profile`:
```sh
DAEDALUS_DIR=/mnt/archive/daedalus
export DAEDALUS_DIR
```
...and the contents of that directory is:
```
└── mainnet
    ├── chain
    │   ├── immutable
    │   ├── ledger
    │   └── volatile
    ├── Logs
    │   └── pub
    ├── Secrets
    ├── tls
    │   ├── client
    │   └── server
    └── wallets
```
_(so not a full home directory, only the data)_

**Comment by danger89** [comment]
> Here is my take on it (a bit more generic). I just modified the `~/.local/bin/daedalus-mainnet` script into this:

Hopefully something like this can be merged upstream. So it's at least way better then now.

**Comment by cleverca22** [comment]
any changes you do to the script will undo at the next update

my personal preference is symlinks:
```
[clever@amd-nixos:~/.local/share/Daedalus]$ ls -lh
total 29K
lrwxrwxrwx 1 clever users 27 Mar 29  2018 mainnet -> /home/clever/dedup/mainnet/
```
that will persist between updates, and seamlessly redirect everything

**Comment by ancoron** [comment]
@cleverca22 Symlinking only works in your case because you kept everything under `/home`. Just look at the bind mount options that the standard version is doing in that start script (`/home` and `/etc`). So it simply can't work for other locations outside of those two directories (also: the symlink will be there, but Daedalus will not be able to follow it, since the nix chroot doesn't allow that.

**Comment by cleverca22** [comment]
ah yeah, the chroot complicates matters, anywhere you move it must still be accessible

**Comment by user1n** [comment]
Came here only to post this https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html#hardlinks
Hoping it helps some of you as it has me.

**Comment by danger89** [comment]
Soft-links and hard-links are all a very bad workaround IMO. 

Why can't I easily change the Daedalus wallet storage location in the GUI? After 5 years!!

**Comment by danger89** [comment]
@danielmain Can this feature be planned in any upcoming releases?

This missing feature is hampering me to use a good decentralized Cardano wallet on my PC (no web-wallets).

**Comment by KIVassilev** [comment]
> You have multiple options now. If you dont need to run a full node, you can simply switch to one of the light wallets: 
I do want to run a full node, just not on my system drive.

**Comment by danger89** [comment]
> with no native solution on the horizon

Yea so this bothers me quite a bit. I looks like there is no focus on Daedalus wallet for the past years as you mentioned. Cardano project is now even focusing in creating their own light wallet. Which is fine as well, but IMO Cardano project/IOHK should definitely focus on Daedalus. Especially on the user ability side, like this kind of issues.

**Comment by danielmain** [comment]
> @danielmain Can this feature be planned in any upcoming releases?
> 
> This missing feature is hampering me to use a good decentralized Cardano wallet on my PC (no web-wallets).

>Yea so this bothers me quite a bit. I looks like there is no focus on Daedalus wallet for the past years as you mentioned. Cardano project is now even focusing in creating their own light wallet. Which is fine as well, but IMO Cardano project/IOHK should definitely focus on Daedalus. Especially on the usability side, like this kind of issues.

@danger89 I created a task for this in our backlog. If our product owner thinks that this a high requested feature, I can imagine creating a PR for it.

**Comment by danielmain** [comment]
Our PO set this as low priority because we have a great workaround described here: https://iohk.zendesk.com/hc/en-us/articles/900004340586-How-to-symlink-Daedalus-chain-folder

Actually my chain folder is in my NAS on my home network and Daedalus works perfectly:
<img width="1046" alt="CleanShot 2022-07-11 at 21 40 18@2x" src="https://user-images.githubusercontent.com/4303099/178345160-29568f90-0b51-454d-a3aa-c757b66ca39f.png">

We might create some option to let users move the chain folder somewhere else but not now.

**Comment by danielmain** [comment]
@tbidne, we have an internal JIRA ticket in our backlog with low priority so that we will not forget it :)
<img width="600" alt="image" src="https://user-images.githubusercontent.com/4303099/178457861-875f6e1a-676b-410d-b2f0-fa347db45179.png">
As soon we start with the implementation I will parte the PR link so you can track it.

> Google something like "daedalus install directory" and hopefully find this issue or that blog post.
Well seems easy to find it:
<img width="973" alt="image" src="https://user-images.githubusercontent.com/4303099/178455335-8046cb2f-68dc-4af6-9130-9e727527e3bd.png">
<img width="867" alt="image" src="https://user-images.githubusercontent.com/4303099/178455417-d9fe85a4-4ba7-48fd-936f-db8eb99ba51a.png">

> Create permanent symlinks.
What do you mean by permanent? symlinks stay as long you do not delete them

@tbidne trust me is not that we don't want to implement it. However, we are very busy with other vital features that are more important than this one. I hope you understand.

**Comment by danger89** [comment]
Also keep in mind that if you already created a symlink before the Daedalus installation (executing: `daedalus-x.x.x-mainnet-x.bin`). You will get an error:

```
+ test -z ''
+ XDG_DATA_HOME=/home/melroy/.local/share
+ export DAEDALUS_DIR=/home/melroy/.local/share/Daedalus/mainnet
+ DAEDALUS_DIR=/home/melroy/.local/share/Daedalus/mainnet
+ mkdir -pv /home/melroy/.local/share/Daedalus/mainnet/Logs/pub
mkdir: cannot create directory '/home/melroy/.local/share/Daedalus/mainnet': File exists
```

Please consider reopening this issue. Since it's still relevant and needs to be fixed.

**Comment by danger89** [comment]
**Update:** I noticed that latest versions of Daedalus is not using that much storage anymore. Because it's consolidating the data. This would mean less storage requirements, correct?
